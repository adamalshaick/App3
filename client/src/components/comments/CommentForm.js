import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import TextAreaFieldGroup from "../common/TextAreaFieldGroup";
import { addComment } from "../../actions/postActions";

class CommentForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      text: ""
    };
  }

  componentWillReceiveProps(newProps) {
    if (newProps.errors) {
      this.setState({ errors: newProps.errors });
    }
  }

  onSubmit = e => {
    e.preventDefault();

    const { user } = this.props.auth;
    const { postId } = this.props;

    const newComment = {
      text: this.state.text,
      name: user.name,
      avatar: user.avatar
    };

    this.props.addComment(postId, newComment);
    this.setState({ text: "" });
  };

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };
  render() {
    return (
      <div className="col-12">
        <form onSubmit={this.onSubmit}>
          <TextAreaFieldGroup
            className="form-control"
            placeholder="Make a comment..."
            name="text"
            value={this.state.text}
            onChange={this.onChange}
            style={{ fontSize: "12px", width: "100%" }}
          />

          <div className="float-right">
            <button type="submit" className="btn btn-outline-primary btn-sm">
              Submit
            </button>
          </div>
        </form>
      </div>
    );
  }
}

CommentForm.propTypes = {
  addPost: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  postId: PropTypes.string.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(
  mapStateToProps,
  { addComment }
)(CommentForm);
