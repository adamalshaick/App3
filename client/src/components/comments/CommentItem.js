import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { deleteComment } from "../../actions/postActions";
import Link from "react-router-dom/Link";

class CommentItem extends Component {
  onDeleteClick(postId, commentId) {
    this.props.deleteComment(postId, commentId);
  }

  render() {
    const { comment, postId, auth } = this.props;
    return (
      <div className=" mb-3">
        <div className="row mt-3">
          <Link to="">
            <img
              style={{ width: "40px" }}
              className="rounded-circle  ml-4"
              src={`../uploads/post_image/${profile.profileImage}`}
              alt=""
            />
          </Link>

          <div className="text-center mt-2 ml-2">{comment.name}</div>

          <div className="col-12">
            <div style={{ fontSize: "0.9rem" }} className="float-left ml-2">
              {comment.text}
            </div>
          </div>

          {comment.user === auth.user.id ? (
            <button
              onClick={this.onDeleteClick.bind(this, postId, comment._id)}
              type="button"
              className="btn btn-outline-danger btn-sm  mr-1"
            >
              Delete
            </button>
          ) : null}
        </div>
      </div>
    );
  }
}

CommentItem.propTypes = {
  deleteComment: PropTypes.func.isRequired,
  comment: PropTypes.object.isRequired,
  postId: PropTypes.string.isRequired,
  auth: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(
  mapStateToProps,
  { deleteComment }
)(CommentItem);
