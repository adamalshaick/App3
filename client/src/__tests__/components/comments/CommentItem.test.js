import React from "react";
import { shallow } from "enzyme";
import { CommentItem } from "../../../components/comments/CommentItem";
import toJson from "enzyme-to-json";

const mockDeleteComment = jest.fn();

describe("render component", () => {
  const Props = {
    comment: {},
    auth: {
      user: {
        id: "id"
      }
    },
    postId: "postId"
  };

  const wrapper = shallow(
    <CommentItem {...Props} deleteComment={mockDeleteComment} />
  );

  it("renders component", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe("current user's comment", () => {
  const PropsWhenIdsMatch = {
    comment: {
      user: "userId"
    },
    auth: {
      user: {
        id: "userId"
      }
    },
    postId: "postId"
  };
  const wrapper = shallow(
    <CommentItem {...PropsWhenIdsMatch} deleteComment={mockDeleteComment} />
  );
  it("shows delete button", () => {
    expect(wrapper.find("#deleteButton").length).toEqual(1);
  });
  it("calls delete fn", () => {
    wrapper.find("#deleteButton").simulate("click", { preventDefault() {} });
    expect(mockDeleteComment.mock.calls.length).toEqual(1);
  });
});

describe("other user's comment", () => {
  const PropsWhenIdsDontMatch = {
    comment: {
      user: "otherUserId"
    },
    auth: {
      user: {
        id: "userId"
      }
    },
    postId: "postId"
  };
  const wrapper = shallow(
    <CommentItem {...PropsWhenIdsDontMatch} deleteComment={mockDeleteComment} />
  );
  it("doesnt show delete button", () => {
    expect(wrapper.find("#deleteButton").length).toEqual(0);
  });
});
