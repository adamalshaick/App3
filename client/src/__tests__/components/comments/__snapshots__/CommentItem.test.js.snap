// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render component renders component 1`] = `
<div
  className=" mb-3"
>
  <div
    className="row mt-3"
  >
    <img
      alt=""
      className="rounded-circle  ml-4"
      style={
        Object {
          "height": "40px",
          "width": "40px",
        }
      }
    />
    <div
      className="text-center mt-2 ml-2"
    />
    <div
      className="col-12"
    >
      <div
        className="float-left ml-2"
        style={
          Object {
            "fontSize": "0.9rem",
          }
        }
      />
    </div>
  </div>
</div>
`;
